File: linkedLists.cpp
#include <iostream>
using namespace std;

// Node structure for the linked list
struct Node {
    int data;         // Data part
    Node* next;       // Pointer to the next node
};

// Function to insert a node at the beginning of the linked list
void insertAtBeginning(Node*& head, int newData) {
    Node* newNode = new Node();  // Allocate new node
    newNode->data = newData;     // Put in the data
    newNode->next = head;        // Make next of new node as head
    head = newNode;              // Move the head to point to the new node
}

// Function to insert a node at the end of the linked list
void insertAtEnd(Node*& head, int newData) {
    Node* newNode = new Node();  // Allocate new node
    newNode->data = newData;     // Put in the data
    newNode->next = nullptr;     // Make the next of new node as null

    // If the linked list is empty, then make the new node the head
    if (head == nullptr) {
        head = newNode;
        return;
    }

    // Else traverse till the last node
    Node* last = head;
    while (last->next != nullptr) {
        last = last->next;
    }

    // Change the next of the last node
    last->next = newNode;
}

// Function to print the linked list
void printList(Node* node) {
    while (node != nullptr) {
        cout << node->data << " -> ";
        node = node->next;
    }
    cout << "NULL" << endl;
}

// Main function to test the linked list operations
int main() {
    Node* head = nullptr;  // Initialize head pointer to nullptr

    // Insert elements at the beginning
    insertAtBeginning(head, 10);
    insertAtBeginning(head, 20);
    insertAtBeginning(head, 30);

    // Insert elements at the end
    insertAtEnd(head, 40);
    insertAtEnd(head, 50);

    // Print the linked list
    cout << "Linked List: ";
    printList(head);

    return 0;
}

File: linkedLists.java
class LinkedList {

    // Node class to represent a node in the linked list
    static class Node {
        int data;      // Data part
        Node next;     // Pointer to the next node

        // Constructor to create a new node
        Node(int d) {
            data = d;
            next = null;
        }
    }

    // Head of the linked list
    Node head;

    // Function to insert a node at the beginning of the linked list
    public void insertAtBeginning(int newData) {
        Node newNode = new Node(newData); // Create a new node
        newNode.next = head;              // Make next of new node as head
        head = newNode;                   // Move the head to point to the new node
    }

    // Function to insert a node at the end of the linked list
    public void insertAtEnd(int newData) {
        Node newNode = new Node(newData); // Create a new node
        if (head == null) {               // If the linked list is empty, make the new node the head
            head = newNode;
            return;
        }

        Node last = head;                 // Traverse to the last node
        while (last.next != null) {
            last = last.next;
        }

        last.next = newNode;              // Change the next of the last node
    }

    // Function to print the linked list
    public void printList() {
        Node currentNode = head;
        while (currentNode != null) {
            System.out.print(currentNode.data + " -> ");
            currentNode = currentNode.next;
        }
        System.out.println("NULL");
    }

    // Main method to test the linked list operations
    public static void main(String[] args) {
        LinkedList list = new LinkedList();

        // Insert elements at the beginning
        list.insertAtBeginning(10);
        list.insertAtBeginning(20);
        list.insertAtBeginning(30);

        // Insert elements at the end
        list.insertAtEnd(40);
        list.insertAtEnd(50);

        // Print the linked list
        System.out.print("Linked List: ");
        list.printList();
    }
}
